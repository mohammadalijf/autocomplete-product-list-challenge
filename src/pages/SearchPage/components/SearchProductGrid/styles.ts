import { makeStyles } from "@material-ui/core";

const useSearchProductGridStyles = makeStyles((theme) => ({
  instructionsLabel: { textAlign: "center" },
}));

export default useSearchProductGridStyles;
