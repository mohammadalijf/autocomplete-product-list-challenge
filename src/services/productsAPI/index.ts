import { parse as csvParser } from "papaparse";
import ProductFilteringWorker from "../../worker/ProductFilteringWorker";
import { IProductQuery } from "./parameters/IProductQuery";
import { ICSVProducts, CSVProductGender } from "./responses/ICSVProducs";

/**
 * fetch array of products from products.csv
 * @returns All the products in products.csv as array of ICSVProduct
 */
async function fetchProducts(): Promise<ICSVProducts[]> {
  const promise = new Promise<ICSVProducts[]>((resolve, reject) => {
    csvParser("/products.csv", {
      download: true,
      header: true,
      delimiter: ",",
      skipEmptyLines: true,
      error: (error) => {
        reject(error);
      },
      complete: (results) => {
        resolve(
          results.data.map<ICSVProducts>((item: any) => ({
            title: item.title,
            gtin: item.gtin,
            gender: Object.values(CSVProductGender).includes(item.gender)
              ? item.gender
              : undefined,
            salePrice: item.sale_price,
            price: item.price,
            imageLink: item.image_link,
            additionalImageLink: item.additional_image_link.split(","),
            onSale: item.sale_price !== item.price,
          }))
        );
      },
    });
  });
  return promise;
}

/**
 * Filters Products base on query on their title or gtin, and gender and onSale
 * @param products initial product sets to apply filters on
 * @param queries query object that may contain query, onSale, gender
 * @param signal abort signal for canceling the thread and process
 * @returns an array of products that passed the filtering
 */
async function searchProducts(
  products: ICSVProducts[],
  queries: IProductQuery,
  signal?: AbortSignal
): Promise<ICSVProducts[]> {
  // init a new worker
  const worker = new ProductFilteringWorker();

  // set up for abort signal
  const abortHandler = () => {
    // kill the worker if got abort signal
    worker.terminate();
  };
  signal?.addEventListener("abort", abortHandler);

  const filteredProduct = await worker.filterProducts(products, queries);
  // filtering is done so remove abort listener and return items
  signal?.removeEventListener("abort", abortHandler);
  return filteredProduct;
}

const productsAPI = { fetchProducts, searchProducts };

export type { ICSVProducts };
export { CSVProductGender };
export default productsAPI;
