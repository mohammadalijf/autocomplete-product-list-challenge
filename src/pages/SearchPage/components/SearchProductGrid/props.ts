import { ICSVProducts } from "../../../../services/productsAPI";

export type SearchProductGridProps = {
  /**
   * list of products to show in grid
   */
  products: ICSVProducts[];
  /**
   * boolean to show instruction label in grid
   */
  showInstructions?: boolean;
  /**
   * error message to show in grid
   */
  error?: string;
  /**
   * loading boolean to show circular progress in grid
   */
  loading?: boolean;
};
