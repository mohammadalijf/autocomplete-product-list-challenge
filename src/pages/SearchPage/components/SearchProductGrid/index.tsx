import { Typography } from "@material-ui/core";
import { FunctionComponent } from "react";
import { SearchProductGridProps } from "./props";
import useSearchProductGridStyles from "./styles";

const SearchProductGrid: FunctionComponent<SearchProductGridProps> = ({
  products,
  showInstructions,
  showEmpty,
  ...props
}) => {
  const classes = useSearchProductGridStyles();
  return (
    <section data-testid="searchProductGrid">
      {showInstructions && (
        <Typography className={classes.instructionsLabel}>
          Start searching for products by typing in the search box
        </Typography>
      )}
      {showEmpty && (
        <Typography className={classes.instructionsLabel}>
          Sorry! we didn't find anything... maybe try for another product?
        </Typography>
      )}
    </section>
  );
};

export default SearchProductGrid;
