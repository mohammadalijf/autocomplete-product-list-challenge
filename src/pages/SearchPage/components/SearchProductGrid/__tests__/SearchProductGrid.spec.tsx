import { cleanup, render } from "@testing-library/react";
import SearchProductGrid from "..";
import {
  fetchProducts,
  NotFoundProductError,
} from "../../../../../reducers/products";
import { ICSVProducts } from "../../../../../services/productsAPI";
import { store } from "../../../../../store";

describe("<SearchProductGrid />", () => {
  afterEach(cleanup);
  let products: ICSVProducts[];

  beforeEach(async () => {
    await store.dispatch(fetchProducts());
    products = store.getState().products.products.slice(0, 2);
  });

  it("renders correctly", async () => {
    let searchProductGrid = render(<SearchProductGrid products={products} />);
    expect(searchProductGrid.asFragment()).toMatchSnapshot();
  });

  it("should show error text when error is passed", () => {
    const error = new NotFoundProductError();
    let searchProductGrid = render(
      <SearchProductGrid products={[]} error={error.message} />
    );
    expect(searchProductGrid.baseElement).toHaveTextContent(error.message);
  });

  it("should show instructions text when showInstructions is passed", () => {
    let searchProductGrid = render(
      <SearchProductGrid products={[]} showInstructions />
    );
    expect(searchProductGrid.baseElement).toHaveTextContent(
      "Start searching for products by typing in the search box"
    );
  });

  it("should show correct product count", () => {
    let searchProductGrid = render(<SearchProductGrid products={products} />);
    expect(searchProductGrid.queryAllByTestId("productCell")).toHaveLength(
      products.length
    );
  });
});
