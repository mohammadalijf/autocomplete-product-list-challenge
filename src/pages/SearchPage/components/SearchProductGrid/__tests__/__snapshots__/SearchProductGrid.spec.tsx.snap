// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchProductGrid /> renders correctly 1`] = `
<DocumentFragment>
  <section
    class="makeStyles-container-1"
    data-testid="searchProductGrid"
  >
    <article
      class="makeStyles-container-4"
      data-testid="productCell"
    >
      <div
        class="makeStyles-imgContainer-8"
      >
        <img
          alt="Weekday THURSDAY Jeans Slim Fit black"
          class="makeStyles-img-9"
          data-testid="productMainImg"
          loading="lazy"
          src="/placeholder.png"
        />
      </div>
      <h6
        class="MuiTypography-root makeStyles-titleLabel-5 MuiTypography-body1"
        data-testid="productTitle"
      >
        Weekday THURSDAY Jeans Slim Fit black
      </h6>
      <div
        class="makeStyles-inforContainer-6"
      >
        <div>
          <span
            class="MuiTypography-root makeStyles-infoTitleLabel-7 MuiTypography-caption"
          >
            GTIN
          </span>
          <span
            class="MuiTypography-root MuiTypography-caption"
            data-testid="productGTIN"
          >
            2001007926858
          </span>
        </div>
        <div>
          <span
            class="MuiTypography-root makeStyles-infoTitleLabel-7 MuiTypography-caption"
          >
            Gender
          </span>
          <span
            class="MuiTypography-root MuiTypography-caption"
            data-testid="productGender"
          >
            female
          </span>
        </div>
        <div>
          <span
            class="MuiTypography-root makeStyles-infoTitleLabel-7 MuiTypography-caption"
          >
            Price
          </span>
          <span
            class="MuiTypography-root MuiTypography-caption"
            data-testid="productPrice"
          >
            39.95 EUR
          </span>
        </div>
        <div>
          <span
            class="MuiTypography-root makeStyles-infoTitleLabel-7 MuiTypography-caption"
          >
            Sale Price
          </span>
          <span
            class="MuiTypography-root MuiTypography-caption"
            data-testid="productSalePrice"
          >
            39.95 EUR
          </span>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
        data-testid="moreImageButton"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          show more images
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </article>
    <article
      class="makeStyles-container-4"
      data-testid="productCell"
    >
      <div
        class="makeStyles-imgContainer-8"
      >
        <img
          alt="Weekday THURSDAY Jeans Slim Fit black"
          class="makeStyles-img-9"
          data-testid="productMainImg"
          loading="lazy"
          src="/placeholder.png"
        />
      </div>
      <h6
        class="MuiTypography-root makeStyles-titleLabel-5 MuiTypography-body1"
        data-testid="productTitle"
      >
        Weekday THURSDAY Jeans Slim Fit black
      </h6>
      <div
        class="makeStyles-inforContainer-6"
      >
        <div>
          <span
            class="MuiTypography-root makeStyles-infoTitleLabel-7 MuiTypography-caption"
          >
            GTIN
          </span>
          <span
            class="MuiTypography-root MuiTypography-caption"
            data-testid="productGTIN"
          >
            2001008772980
          </span>
        </div>
        <div>
          <span
            class="MuiTypography-root makeStyles-infoTitleLabel-7 MuiTypography-caption"
          >
            Gender
          </span>
          <span
            class="MuiTypography-root MuiTypography-caption"
            data-testid="productGender"
          >
            female
          </span>
        </div>
        <div>
          <span
            class="MuiTypography-root makeStyles-infoTitleLabel-7 MuiTypography-caption"
          >
            Price
          </span>
          <span
            class="MuiTypography-root MuiTypography-caption"
            data-testid="productPrice"
          >
            39.95 EUR
          </span>
        </div>
        <div>
          <span
            class="MuiTypography-root makeStyles-infoTitleLabel-7 MuiTypography-caption"
          >
            Sale Price
          </span>
          <span
            class="MuiTypography-root MuiTypography-caption"
            data-testid="productSalePrice"
          >
            39.95 EUR
          </span>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
        data-testid="moreImageButton"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          show more images
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </article>
  </section>
</DocumentFragment>
`;
